en/ros2/px4_sdk.md

```diff
@@ -0,0 +1,336 @@
+
+### Setpoint Types
+
+A mode can choose its setpoint type(s) it wants to use to control the vehicle.
+The used types also define the compatibility with different vehicle types.
+
+The following sections provide a list of commonly used setpoint types.
+You can also add your own type by adding a class that inherits from `px4_sdk::SetpointBase`, sets the configuration flags according to what the setpoint requires, and then publishes any topic containing a setpoint.
+
+#### Smooth Position
+
+TODO
+
+#### Direct Actuator Control
+
+Actuators can be directly controlled using the [px4_sdk::DirectActuatorsSetpointType](https://github.com/PX4/px4_sdk/blob/main/px4_sdk_cpp/include/px4_sdk/control/setpoint_types/direct_actuators.hpp) setpoint type.
+Motors and servos can be set independently. Be aware that the assignment is vehicle and setup-specific.
+For example to control a quadrotor, you need to set the first 4 motors according to its [output configuration](../concept/control_allocation.md).
+
+:::note
+If you want to control an actuator that does not control the vehicle's motion, but for example a payload servo, see [below](#controlling-an-independent-actuator-servo).
+:::
+
+### Controlling an Independent Actuator/Servo
+
+If you want to control an independent actuator (a servo), follow these steps:
+
+1. [Configure the output](../payloads/#generic-actuator-control-with-mavlink)
+2. Create an instance of [px4_sdk::OffboardActuatorControls](https://github.com/PX4/px4_sdk/blob/main/px4_sdk_cpp/include/px4_sdk/control/offboard_actuators.hpp) in the constructor of your mode
+3. Call the `set` method to control the actuator(s). This can be done independently of any active setpoints.
+
+### Telemetry
+
+Telemetry, such as local or global position estimates can be found under [px4_sdk/odometry](https://github.com/PX4/px4_sdk/tree/main/px4_sdk_cpp/include/px4_sdk/odometry).
+
+### Failsafes and Mode Requirements
+
+Each mode has a set of requirement flags. These are generally automatically set, depending on which objects are used within the context of a mode. For example when adding manual control input with
+
+```cpp
+_manual_control_input = std::make_shared<px4_sdk::ManualControlInput>(*this);
+```
+
+the requirement flag for manual control gets set.
+Specifically, setting a flag has the following consequences in PX4, if the condition is not met:
+
+- arming is not allowed, while the mode is selected
+- when already armed, the mode cannot be selected
+- when armed and the mode is selected, the relevant failsafe is triggered (e.g. RC loss for the manual control requirement).
+  Check the [safety page](../config/safety.md) for how to configure failsafe behavior.
+  A failsafe is also triggered when the mode crashes or becomes unresponsive while it is selected.
+
+It is possible to manually update any mode requirement after the mode is registered. For example to add home position as requirement:
+
+```cpp
+modeRequirements().home_position = true;
```

Comment:
Can we list all the requirements you can set, or link to how you can find them?

The example above of manual mode and what gets set is awesome. Would be good to expand out the other modes and their implicit requirements. Why I'm interested here is that it's good to make it easy for people to understand what they can do, and when they might be able to do it.

The trade off is time to document and time to maintain. So you might alternatively link to source that shows the options, or source code that shows how you can easily list the requirements (all of them) that the mode currently has.